src/client/Client.cc

```diff
@@ -12806,13 +12806,15 @@ int Client::_setxattr_check_data_pool(string& name, string& value, const OSDMap
     int64_t pool;
     try {
       pool = boost::lexical_cast<unsigned>(tmp);
-      if (!osdmap->have_pg_pool(pool))
-	return -CEPHFS_ENOENT;
-    } catch (boost::bad_lexical_cast const&) {
-      pool = osdmap->lookup_pg_pool_name(tmp);
-      if (pool < 0) {
-	return -CEPHFS_ENOENT;
+      if (osdmap->have_pg_pool(pool)) {
+          return 0;
       }
+    } catch (boost::bad_lexical_cast const&) {
+        ldout(cct, 15) << __func__ << ": pool in layout is not a pool id: " << tmp << dendl;
+    }
+    pool = osdmap->lookup_pg_pool_name(tmp);
+    if (pool < 0) {
+        return -CEPHFS_ENOENT;
```

Comment:
Could you explain why the old code didn't work ?

Reply:
If pool name is a positive integer and can be cast to unsigned. layout.pool is set to this pool name.
And then, use this name as id to ckeck the pool in osdmap.
If there is no pool with this id, it will return -ENOENT and no longer check as a pool name.

Reply:
If the `layout.pool` is set to pool name, then the `pool = boost::lexical_cast<unsigned>(tmp);` should throw a exception and will be catched, then in the catch it was checking with the pool name.

Reply:
If pool name is a positive integer，the name can be cast to unsigned. no exception will be thrown.

Reply:
Do you mean the pool name will be something like `23134`, which only includes digit numbers, but this is not a pool id ?

Reply:
yes

Reply:
Okay. Make sense. Please explain this in the commit comment.

Reply:
I'm being paranoid, but what if there are 2 pools, one with the pool name as `"12345"` and another pool with id as `12345`.